### PR TITLE
Support for fallback keys

### PR DESCRIPTION
**Pull request by neilalexander**
Backports support for fallback keys from Harmony, which should make E2EE more reliable in the face of OTK exhaustion. 

Signed-off-by: Neil Alexander <git@neilalexander.dev>

### CODE REVIEW

File: userapi/internal/key_api.go [hunk 1]
```diff
@@ -768,7 +790,30 @@ func (a *UserInternalAPI) uploadOneTimeKeys(ctx context.Context, req *api.Perfor
...
+		for keyIDWithAlgo := range key.KeyJSON {
+			keyIDsWithAlgorithms[i] = keyIDWithAlgo
+			i++
+		}
+		if err := a.KeyDatabase.DeleteFallbackKeys(ctx, req.UserID, req.DeviceID); err != nil {
+			res.KeyError(req.UserID, req.DeviceID, &api.KeyError{
+				Err: fmt.Sprintf("%s device %s : failed to clear fallback keys: %s", req.UserID, req.DeviceID, err.Error()),
+			})
+			continue
+		}
```
Comment (S7evinK) [inline userapi/internal/key_api.go:806]: Not totally sure why we need to clear existing fallback keys, but doing that in this loop seems odd.

We clear the keys, store a new one, clear the keys, store a new one, etc. Don't we end up with just one key?

**Review by S7evinK — CHANGES_REQUESTED** [review]
Looks sane to me, just the dupe indexes and the strange clearing of the fallback keys in the loop which need some clarification.

File: userapi/internal/key_api.go (hunk 1 above)
Comment (neilalexander) [reply userapi/internal/key_api.go]: Fixed this.

**Review by S7evinK — APPROVED** [review]
LGTM now! Thanks!

**Pull request merged**